<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Using khmer on real data &mdash; zaok 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="zaok 0.1 documentation" href="index.html" />
    <link rel="next" title="Perspectives and considerations" href="considerations.html" />
    <link rel="prev" title="Summarizing the basics" href="summarizing-khmer-basics.html" /> 
  </head>
  <body role="document">
      <div class="header" role="banner"><h1 class="heading"><a href="index.html">
          <span>zaok 0.1 documentation</span></a></h1>
        <h2 class="heading"><span>Using khmer on real data</span></h2>
      </div>
      <div class="topnav" role="navigation" aria-label="top navigation">
      
        <p>
        «&#160;&#160;<a href="summarizing-khmer-basics.html">Summarizing the basics</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="considerations.html">Perspectives and considerations</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="using-khmer-on-real-data">
<h1>Using khmer on real data<a class="headerlink" href="#using-khmer-on-real-data" title="Permalink to this headline">¶</a></h1>
<p>khmer contains a set of command-line scripts that wrap the library code.
These scripts can serve a variety of purposes; the main ones are
k-mer abundance distribution analysis, k-mer abundance trimming,
digital normalization, and partitioning.  There are also a set of
scripts for data handling (e.g. read interleaving) that are just useful
to have around.</p>
<div class="section" id="some-background-philosophy">
<h2>Some background philosophy<a class="headerlink" href="#some-background-philosophy" title="Permalink to this headline">¶</a></h2>
<p>We have largely shied away from actually producing a final result, and
instead focused on building tools and utilites that manipulate data
sets between generation and analysis.  In essence, we try to act as a
filtering and forensics tool.  There are several reasons for this:</p>
<ul class="simple">
<li>flexibility: providing tools that can be mixed and matched is more
flexible than providing a single tool.</li>
<li>format: consuming and producing common read formats is much easier
to get right than producing fast-evolving &#8220;final&#8221; output formats.</li>
<li>simplicity: many simple tools can still yield a complex result, while
being simple to run and easy to understand (at least individually).</li>
<li>correctness: simplicity and correctness are linked ;).</li>
<li>scalability: simple tools can often be made to run quite fast.</li>
<li>streaming: many khmer tools support streaming input &amp; output.</li>
</ul>
<p>So, most of the tools and approaches described below act to slice and
dice data, with the goal of feeding the data into other programs.</p>
<p>The downsides to this approach are:</p>
<ul class="simple">
<li>bioinformatics formats still change frequently enough that keeping
khmer working with them is a chore.</li>
<li>it adds to the complexity of the process.  Many biologists would like
a tool that simply <em>does the job</em>, and khmer is most emphatically not
that tool.</li>
<li>people may (rightly or wrongly) blame khmer when their results are
not good.  For example, people who apply digital normalization
sometimes expect that they can continue to use their downstream
tools with unchanged parameters, even though diginorm alters the
coverage distribution; this is sometimes true, and sometimes not.</li>
</ul>
</div>
<div class="section" id="k-mer-abundance-distribution-analysis">
<h2>K-mer abundance distribution analysis<a class="headerlink" href="#k-mer-abundance-distribution-analysis" title="Permalink to this headline">¶</a></h2>
<p>With khmer, it is straightforward to compute k-mer abundance
histograms (equiv.  k-mer spectra) and related values from very large
data sets.  The procedure is as follows:</p>
<ol class="arabic simple">
<li>Create a countgraph from all of the data (<code class="docutils literal"><span class="pre">load-into-counting.py</span></code>).</li>
<li>Count the k-mers in a query data set using the counts in the countgraph;</li>
<li>Summarize the counts (e.g. <code class="docutils literal"><span class="pre">abundance-dist.py</span></code>).</li>
</ol>
<p>The most straightforward use for this is just computing the k-mer
abundance spectra; this can be done in one step for single files with
<code class="docutils literal"><span class="pre">abundance-dist-single.py</span></code>, but for multiple data files you&#8217;ll need to
use <code class="docutils literal"><span class="pre">load-into-counting.py</span></code> and concatenate the files into <code class="docutils literal"><span class="pre">abundance-dist.py</span></code>.</p>
<p>There are several utilities, including read coverage distribution
analysis (&#64;&#64;).</p>
</div>
<div class="section" id="k-mer-abundance-trimming">
<h2>K-mer abundance trimming<a class="headerlink" href="#k-mer-abundance-trimming" title="Permalink to this headline">¶</a></h2>
<p>khmer is particularly good at removing low-abundance k-mers from data
sets, for the purpose of removing errors and decreasing memory
requirements (see XXX).  Typically this is done on a per-read basis,
where each read (or pair of reads) is examined in the context of
a k-mer database, and then the read(s) are trimmed at a given position.</p>
<p>We have two methods for doing this, non-streaming and streaming.</p>
<p>The non-streaming procedure is:</p>
<ol class="arabic simple">
<li>Create a countgraph from all of the data.</li>
<li>For each data set</li>
</ol>
<p>Here an important consideration is that &#8220;hard&#8221; k-mer abundance
trimming (where you eliminate any k-mer with an abundance below a
threshold) needs to be done in the context of the entire data set.
&#8220;Soft&#8221; or variable-coverage k-mer abundance trimming can be done
on each read data set individually, but should then be re-done on
the whole data set.</p>
<p>Note, you can trim high-abundance k-mers off of reads with
<code class="docutils literal"><span class="pre">sandbox/trim-below-abund.py</span></code>; this helps reduce and remove repeats.</p>
</div>
<div class="section" id="digital-normalization">
<h2>Digital normalization<a class="headerlink" href="#digital-normalization" title="Permalink to this headline">¶</a></h2>
<p>&#64;&#64;can be done successively</p>
</div>
<div class="section" id="partitioning">
<h2>Partitioning<a class="headerlink" href="#partitioning" title="Permalink to this headline">¶</a></h2>
<p>We don&#8217;t really recommend using this at the moment, but here&#8217;s roughly
how it works:</p>
<p>...</p>
</div>
<div class="section" id="recipes">
<h2>Recipes<a class="headerlink" href="#recipes" title="Permalink to this headline">¶</a></h2>
<p>We have a few recipes showing how to use various khmer scripts to do
analyses that might be of interest &#8211; documents are <a class="reference external" href="https://khmer-recipes.readthedocs.io/en/latest/">here</a>, source is <a class="reference external" href="https://github.com/dib-lab/khmer-recipes">here</a>.</p>
</div>
<div class="section" id="protocols">
<h2>Protocols<a class="headerlink" href="#protocols" title="Permalink to this headline">¶</a></h2>
<p>We have a pair of protocols that explain how to go from raw data to an
assembly, for RNAseq and metagenome data.  The latest version is in
the <code class="docutils literal"><span class="pre">ctb</span></code> branch; the documents are <a class="reference external" href="http://khmer-protocols.readthedocs.io/en/ctb/">here</a>, and the source is
<a class="reference external" href="https://github.com/dib-lab/khmer-protocols">here</a>.</p>
<p>Note: the protocols serve as an important part of the release checklist,
because they verify that khmer works as part of a larger set of tools.</p>
</div>
<div class="section" id="choosing-parameters">
<h2>Choosing parameters<a class="headerlink" href="#choosing-parameters" title="Permalink to this headline">¶</a></h2>
<div class="section" id="memory-parameters-and-where-they-matter">
<h3>Memory parameters, and where they matter<a class="headerlink" href="#memory-parameters-and-where-they-matter" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="k-mer-sizes-and-k-mer-size-limitations">
<h3>K-mer sizes, and k-mer size limitations<a class="headerlink" href="#k-mer-sizes-and-k-mer-size-limitations" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="input-and-output-formats">
<h3>Input and output formats<a class="headerlink" href="#input-and-output-formats" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


      </div>
      <div class="bottomnav" role="navigation" aria-label="bottom navigation">
      
        <p>
        «&#160;&#160;<a href="summarizing-khmer-basics.html">Summarizing the basics</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="considerations.html">Perspectives and considerations</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, C. Titus Brown.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.4.
    </div>
  </body>
</html>